network/src/topology.rs

```diff
@@ -0,0 +1,492 @@
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+// Network crawler:
+// Start a crawler task (similar to the peers task) which updates state. Only one peer would be
+// connected at a time to start and would be queried for peers. It would then select on peer at
+// random to continue the crawl.
+//
+// Q: extend the network protocol to include statistics or node metadata?
+// Q: when to perform centrality computation?
+
+use std::{
+    cmp::Ordering,
+    collections::{BTreeMap, HashSet},
+    hash::{Hash, Hasher},
+    net::SocketAddr,
+    ops::Sub,
+};
+
+use chrono::{DateTime, Utc};
+use nalgebra::{DMatrix, DVector, SymmetricEigen};
+use parking_lot::RwLock;
+
+// Purges connections that haven't been seen within this time (in hours).
+const STALE_CUTOFF: i64 = 4;
+
+/// A connection between two peers.
+///
+/// Implements `partialEq` and `Hash` so that the `source`-`target` order has no impact on equality
+/// (since connections are directionless). The timestamp is also not included in the comparison.
+#[derive(Debug, Eq, Copy, Clone)]
+pub struct Connection {
+    /// One side of the connection.
+    pub source: SocketAddr,
+    /// The other side of the connection.
+    pub target: SocketAddr,
+    /// The last time this peer was seen by the crawler (used determine which connections are
+    /// likely stale).
+    last_seen: DateTime<Utc>,
+}
+
+impl PartialEq for Connection {
+    fn eq(&self, other: &Self) -> bool {
+        let (a, b) = (self.source, self.target);
+        let (c, d) = (other.source, other.target);
+
+        a == d && b == c || a == c && b == d
+    }
+}
+
+impl Hash for Connection {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let (a, b) = (self.source, self.target);
+
+        match a.cmp(&b) {
+            Ordering::Greater => {
+                b.hash(state);
+                a.hash(state);
+            }
+
+            _ => {
+                a.hash(state);
+                b.hash(state);
+            }
+        }
+    }
+}
+
+impl Connection {
+    fn new(source: SocketAddr, target: SocketAddr) -> Self {
+        Connection {
+            source,
+            target,
+            last_seen: Utc::now(),
+        }
+    }
+}
+
+/// Keeps track of crawled peers and their connections.
+#[derive(Debug, Default)]
+pub struct NetworkTopology {
```

Comment:
```suggestion
pub struct KnownNetwork {
```

Technically, this object doesn't contain information on the topology of the network (it can be derived from it), but the network itself.